src/density_methods.jl

```diff
@@ -143,20 +143,24 @@ end
 function atomic_density_superposition(basis::PlaneWaveBasis{T},
                                       form_factors::IdDict{Tuple{Int,T},T};
                                       coefficients=ones(T, length(basis.model.atoms))
-                                      )::Array{T,3} where {T}
```

Comment:
If I recall correctly we did this to avoid a type stablity / help the compiler to infer types. Best to check this in case you have not yet done so (e.g. with Cthulu.jl)
@azadoks do you remember?

Reply:
I vaguely remember having some type stability issues here, but explicitly checking it would be good.

Reply:
OK I will try to investigate it. But have `Array{T,3}` would force the returned density to be on CPU I think.

Reply:
> But have Array{T,3} would force the returned density to be on CPU I think.

Absolutely. If we really need the annotation, in that case we would have to be a bit smarter (e.g. by introducing some type assert or annotation where the type is derived from the array type stored in the PlaneWaveBasis).

Reply:
No, I have not checked, but there is indeed a type instability introduce by this PR due to

```
ρ = to_device(basis.architecture, ρ_cpu)
```

Because it is not possible to infer the `architecture` based on `basis` alone. I think we should include `architecture` in the type parameter of `PlanewaveBasis`. This does not affect the run time of the function itself, though. Is this function a performance critical part and get called lots of times? Otherwise, I think instability does not harm too much.

I checked that the original version did not to have type instability for the output type in the first place anyway, so there is no need to have the `Array` (@azadoks do you remember any detail?) However, there are lots of allocations when called - I think the function itself can be optimised a bit.

Reply:
Maybe it can introduce issue with types other than Float64? I am not sure...

Reply:
Sorry it took me a bit to follow up with this. So it actually was two type instabilities. The one you found @zhubonan, but also one we missed earlier and which in fact caused us to put the explicit return type. With that fixed the return type can also be removed.